src/H5Dint.c

```diff
@@ -3697,11 +3697,15 @@ H5D_get_create_plist(const H5D_t *dset)
     ret_value = new_dcpl_id;
 
 done:
-    if (ret_value < 0)
+    if (ret_value < 0) {
         if (new_dcpl_id > 0)
             if (H5I_dec_app_ref(new_dcpl_id) < 0)
                 HDONE_ERROR(H5E_DATASET, H5E_CANTDEC, FAIL, "unable to close temporary object");
 
+        if (copied_fill.type && (H5T_close_real(copied_fill.type) < 0))
+            HDONE_ERROR(H5E_DATASET, H5E_CANTFREE, FAIL, "Can't free temporary datatype");
```

Comment:
Is this error report redundant, since the call has already failed at this point?

Reply:
One could make the argument that this pattern of reporting errors after the `done:` section in the library when a function has already failed is redundant, but sometimes the error messages can be useful and at least it's uniform so developers don't need to remember special rules about when to report an error or not.